sympy/ntheory/residue_ntheory.py

```diff
@@ -1196,8 +1195,17 @@ def _discrete_log_pollard_rho(n, a, b, order=None, retries=10, rseed=None):
 
             if xa == xb:
                 r = (ba - bb) % order
-                if r != 0:
-                    return mod_inverse(r, order) * (ab - aa) % order
+                if not isprime(order):
```

Comment:
It is not necessary to check whether `order` is prime or not, or whether `r` is 0 or not. `mod_inverse` will work equally well in each case. It will raise `ValueError` if the inverse does not exist, and that should be caught in a try-except block. I would try something like this:

    try:
        e = mod_inverse(r, order) * (ab - aa) % order
        if (pow(b, e, n) - a) % n == 0:
            return e
    except ValueError:
        pass

This code should cover both exceptional cases:  `e` cannot be computed (`mod_inverse` will raise `ValueError`), or it is not a discrete logarithm (`(pow(b, e, n) - a) % n != 0`).

Reply:
right I should have used this only.